File: examples/java/operations/steps/process-data.java
import java.util.Map;
import software.amazon.lambda.durable.DurableContext;
import software.amazon.lambda.durable.DurableHandler;
import software.amazon.lambda.durable.StepContext;
import software.amazon.lambda.durable.config.StepConfig;
import software.amazon.lambda.durable.config.StepSemantics;
import software.amazon.lambda.durable.retry.RetryStrategies;

public class ProcessDataExample extends DurableHandler<Map<String, Object>, Map<String, Object>> {
    @Override
    public Map<String, Object> handleRequest(Map<String, Object> event, DurableContext context) {
        StepConfig config = StepConfig.builder()
            .retryStrategy(RetryStrategies.exponentialBackoff(
                3,
                java.time.Duration.ofSeconds(1),
                java.time.Duration.ofMinutes(5),
                2.0,
                software.amazon.lambda.durable.retry.JitterStrategy.FULL))
            .semantics(StepSemantics.AT_LEAST_ONCE_PER_RETRY)
            .build();

        Map<String, Object> result = context.step("process_data", Map.class,
            (StepContext ctx) -> Map.of("processed", event.get("data"), "status", "completed"),
            config);

        return result;
    }
}

File: examples/csharp/operations/steps/process-data.cs
using Amazon.Lambda.Core;
using Amazon.Lambda.DurableExecution;

public class ProcessDataExample
{
    public Task<DurableExecutionInvocationOutput> Handler(
        DurableExecutionInvocationInput input, ILambdaContext context)
        => DurableFunction.WrapAsync<DataEvent, Processed>(Workflow, input, context);

    private async Task<Processed> Workflow(DataEvent input, IDurableContext ctx)
    {
        var config = new StepConfig
        {
            RetryStrategy = RetryStrategy.Exponential(maxAttempts: 3),
            Semantics = StepSemantics.AtLeastOncePerRetry,
        };

        Processed result = await ctx.StepAsync(
            async (_, _) => ProcessData(input.Data),
            name: "process_data",
            config: config);
        return result;
    }

    private static Processed ProcessData(string data) => new(data, Status: "completed");
}

public record DataEvent(string Data);
public record Processed(string Data, string Status);
